Add support for `command`/`args` fields in `container` components

rm3l opened the pull request:
**What type of PR is this:**
/kind bug

**What does this PR do / why we need it:**
As discussed in https://github.com/redhat-developer/odo/issues/5648#issuecomment-1126009171, the goal of this PR is to support `command` and `args` fields defined in container components by leveraging the fact that `odo dev` is a long-running client-side command.
At this time, Supervisord requires to override the container entrypoint, which is not compliant with the [Devfile specification](https://devfile.io/docs/devfile/2.2.0/user-guide/api-reference/). Per the [Devfile specification](https://devfile.io/docs/devfile/2.2.0/user-guide/api-reference/), we should use any `command` or `args` defined in the container component, or whatever command or entrypoint is available in the container image.
So the approach implemented here is to no longer run containers with Supervisord.
Instead, the pod is now created without the `copy-supervisord` init container, and the container component is created either with the `command` or `args` defined in the Devfile ~, or by not specifying any command, which means Kubernetes/OpenShift will use whatever is defined in the image.~
Then subsequent `run` or `debug` Devfile commands are executed right away in the running container, in a separate goroutine:
- the command process PID is stored in a dedicated file in the container, under `/opt/odo/.odo_devfile_cmd_*.pid`
- the goroutine spawned can be stopped by killing that remote process in the container, which stops the command execution streaming connection to the container
- the command status can be determined by inspecting this PID file

At the moment, most container images used in Devfile stacks define terminating commands, which requires defining explicit commands and/or args if we want to use them with the changes here.
See https://github.com/devfile/registry/pull/102 and https://github.com/devfile/api/issues/679

Because https://github.com/devfile/registry/pull/102 is not merged yet on the Devfile side and to avoid being blocked by this, we decided [1] to implement a temporary hack that sets the container command to `tail -f /dev/null` if none was explicitly defined.

[1] https://github.com/redhat-developer/odo/pull/5768#issuecomment-1147190409

**Which issue(s) this PR fixes:**
Fixes #5648 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**
`odo dev` should work as before, except that:
- Supervisord should no longer be there in the pod.
- the container command should be used as is if defined in the Devfile
- the container command should be overridden with `tail -f /dev/null` if not set in the Devfile (purpose of the temporary hack depicted above)

rm3l commented:
@feloy As discussed yesterday, I've gone ahead and managed to drop (in [e481eff](https://github.com/redhat-developer/odo/pull/5768/commits/e481eff577a4578dc14121972a0767859bfa7c75)) the `pkg/command` package, now relying on `pkg/libdevfile` and `pkg/remotecmd` as much as possible. This way, command handling is in a single place.
Please take another look when you get a chance. Thanks.

dharmit commented:
> To ensure a smooth transition to this new mode, this PR introduces a new `--no-supervisord` flag to `odo dev`.
> The point behind this is mainly for testing purposes, as I wanted to make sure all the existing test cases pass with or without Supervisord.
> This experimental flag is opt-in for now, i.e., it defaults to `false`.

At the moment, the main branch is used to build alpha/beta releases. Putting something in experimental mode for alpha/beta releases, where we have the license/freedom to break stuff, is redundant IMO. We got to be more aggressive/liberal than that, IMHO.

rm3l commented:
> > To ensure a smooth transition to this new mode, this PR introduces a new `--no-supervisord` flag to `odo dev`.
> > The point behind this is mainly for testing purposes, as I wanted to make sure all the existing test cases pass with or without Supervisord.
> > This experimental flag is opt-in for now, i.e., it defaults to `false`.
> 
> At the moment, the main branch is used to build alpha/beta releases. Putting something in experimental mode for alpha/beta releases, where we have the license/freedom to break stuff, is redundant IMO. We got to be more aggressive/liberal than that, IMHO.

Yes, that's a good point - thanks @dharmit. This is experimental from an internal perspective only. This flag is expected to be removed in a near future, so we won't need to support both modes.
But since this changes a lot of things, I wanted to be sure about test coverage first. That's why I intend to submit a separate PR once this one is merged, to switch to this new mode (and clean up the tests).

feloy commented:
> > > To ensure a smooth transition to this new mode, this PR introduces a new `--no-supervisord` flag to `odo dev`.
> > > The point behind this is mainly for testing purposes, as I wanted to make sure all the existing test cases pass with or without Supervisord.
> > > This experimental flag is opt-in for now, i.e., it defaults to `false`.
> > 
> > 
> > At the moment, the main branch is used to build alpha/beta releases. Putting something in experimental mode for alpha/beta releases, where we have the license/freedom to break stuff, is redundant IMO. We got to be more aggressive/liberal than that, IMHO.
> 
> Yes, that's a good point - thanks @dharmit. This is experimental from an internal perspective only. This flag is expected to be removed in a near future, so we won't need to support both modes. But since this changes a lot of things, I wanted to be sure about test coverage first. That's why I intend to submit a separate PR once this one is merged, to switch to this new mode (and clean up the tests).

Also, as I understand, the mode without supervisord will be incompatible with the exitsing devfiles without commands defined in the devfile (and the supervisord mode will break the future devfiles with specified commands). I think it would be nice to have the two modes to manage the transition around this breaking change

rm3l commented:
> > > > To ensure a smooth transition to this new mode, this PR introduces a new `--no-supervisord` flag to `odo dev`.
> > > > The point behind this is mainly for testing purposes, as I wanted to make sure all the existing test cases pass with or without Supervisord.
> > > > This experimental flag is opt-in for now, i.e., it defaults to `false`.
> > > 
> > > 
> > > At the moment, the main branch is used to build alpha/beta releases. Putting something in experimental mode for alpha/beta releases, where we have the license/freedom to break stuff, is redundant IMO. We got to be more aggressive/liberal than that, IMHO.
> > 
> > 
> > Yes, that's a good point - thanks @dharmit. This is experimental from an internal perspective only. This flag is expected to be removed in a near future, so we won't need to support both modes. But since this changes a lot of things, I wanted to be sure about test coverage first. That's why I intend to submit a separate PR once this one is merged, to switch to this new mode (and clean up the tests).
> 
> Also, as I understand, the mode without supervisord will be incompatible with the exitsing devfiles without commands defined in the devfile (and the supervisord mode will break the future devfiles with specified commands). I think it would be nice to have the two modes to manage the transition around this breaking change

Yes, that's right; the `--no-supervisord` command is not compatible with most of the existing Devfiles without explicitly adding `command` or `args` to the container component, unless its image starts a non-terminating command by default.
Actually, the current Supervisord mode remains compatible even with specified `command`/`args`, because since #5639, we are always overriding the container entrypoint with Supervisord. But that does not meet the Devfile specification.
But I agree, we would need to manage the transition around this.

dharmit commented:
> > Also, as I understand, the mode without supervisord will be incompatible with the exitsing devfiles without commands defined in the devfile (and the supervisord mode will break the future devfiles with specified commands). I think it would be nice to have the two modes to manage the transition around this breaking change
> 
> Yes, that's right; the `--no-supervisord` command is not compatible with most of the existing Devfiles without explicitly adding `command` or `args` to the container component, unless its image starts a non-terminating command by default.

Which devfiles are we exactly referring to? The ones in the odo test suite which are modified in this PR? Or the ones maintained on the [registry](https://registry.devfile.io/viewer)?

kadel commented:
> > > Also, as I understand, the mode without supervisord will be incompatible with the exitsing devfiles without commands defined in the devfile (and the supervisord mode will break the future devfiles with specified commands). I think it would be nice to have the two modes to manage the transition around this breaking change
> > 
> > 
> > Yes, that's right; the `--no-supervisord` command is not compatible with most of the existing Devfiles without explicitly adding `command` or `args` to the container component, unless its image starts a non-terminating command by default.
> 
> Which devfiles are we exactly referring to? The ones in the odo test suite which are modified in this PR? Or the ones maintained on the [registry](https://registry.devfile.io/viewer)?
> 
> **EDIT**: Never mind, I found [devfile/registry#102](https://github.com/devfile/registry/pull/102) which addresses this on Devfile side. My question in this case is that when would we be able to default to `--no-supervisord`? At the moment, that PR by Mario is blocked on tooling created atop odo 2.5.0. Unless odo 3.0.0 is GA, that tooling won't/can't update to use odo v3. If 3.0.0 GA were to be blocked on `--no-supervisord` we would have a deadlock like below, IMO:
> 
> ```
> Registry PR 102<-----------no-supervisord<--------IDE plugins
>        |                                               ^
>        |                                               |
>        |                                               |
>        +-----------------------------------------------+
> ```

Dharmit has a good point here. Due to the mess that we created with the original `command/args` implementation, we won't be able to default to `--no-supervisord` until https://github.com/devfile/registry/pull/102 is merged. 
Which could take some time :-(  and I would like for odo to default  `--no-supervisord` and remove the old code with `initContainer` sooner than this. Mantaining booth options is going to be real pain as we keep implementing new features. 

What if we do a small temporary hack: if `command` is not defined in devfile odo would default to `command: ["tail", "-f", "/dev/null"]`. This would be only a temporary solution to allow odo to stop using SupervisordD by default and not have to wait for https://github.com/devfile/registry/pull/102 to get merged. 
It will be much easier to maintain this "hack", which will be just a few lines of code, than maintaining both implementation (with and without the supervisord)

rm3l commented:
> > > > Also, as I understand, the mode without supervisord will be incompatible with the exitsing devfiles without commands defined in the devfile (and the supervisord mode will break the future devfiles with specified commands). I think it would be nice to have the two modes to manage the transition around this breaking change
> > > 
> > > 
> > > Yes, that's right; the `--no-supervisord` command is not compatible with most of the existing Devfiles without explicitly adding `command` or `args` to the container component, unless its image starts a non-terminating command by default.
> > 
> > 
> > Which devfiles are we exactly referring to? The ones in the odo test suite which are modified in this PR? Or the ones maintained on the [registry](https://registry.devfile.io/viewer)?
> > **EDIT**: Never mind, I found [devfile/registry#102](https://github.com/devfile/registry/pull/102) which addresses this on Devfile side. My question in this case is that when would we be able to default to `--no-supervisord`? At the moment, that PR by Mario is blocked on tooling created atop odo 2.5.0. Unless odo 3.0.0 is GA, that tooling won't/can't update to use odo v3. If 3.0.0 GA were to be blocked on `--no-supervisord` we would have a deadlock like below, IMO:
> > ```
> > Registry PR 102<-----------no-supervisord<--------IDE plugins
> >        |                                               ^
> >        |                                               |
> >        |                                               |
> >        +-----------------------------------------------+
> > ```
> 
> Dharmit has a good point here. Due to the mess that we created with the original `command/args` implementation, we won't be able to default to `--no-supervisord` until [devfile/registry#102](https://github.com/devfile/registry/pull/102) is merged. Which could take some time :-( and I would like for odo to default `--no-supervisord` and remove the old code with `initContainer` sooner than this. Mantaining booth options is going to be real pain as we keep implementing new features.
> 
Okay, I understand. I thought https://github.com/devfile/registry/pull/102 would get merged sooner than that. So we are still waiting for most users to use odo >= 2.5.1, right?

> What if we do a small temporary hack: if `command` is not defined in devfile odo would default to `command: ["tail", "-f", "/dev/null"]`. This would be only a temporary solution to allow odo to stop using SupervisordD by default and not have to wait for [devfile/registry#102](https://github.com/devfile/registry/pull/102) to get merged. It will be much easier to maintain this "hack", which will be just a few lines of code, than maintaining both implementation (with and without the supervisors)

Makes sense as a hack. I'll add this and default to `--no-supervisord`. I'll also document this behavior in the `odo dev` documentation, so users are aware of this.

dharmit reviewed (changes requested):
First pass of PR review. Stopping because I can't seem to create a component successfully when I use `odo dev --no-supervisord` for the code at https://github.com/odo-devfiles/nodejs-ex.

dharmit commented on pkg/remotecmd/interface.go:
```diff
@@ -0,0 +1,51 @@
+// in Kubernetes/OpenShift containers, depending on the selected strategy (with or without Supervisord for the moment).
+// See GetRemoteProcessHandlerInstance for retrieving concrete instances of RemoteProcessHandler
+type RemoteProcessHandler interface {
+
```
I got this answered while reading the code, but I would expect it to be a part of this interface because it's providing methods to a bunch of things for a process except getting the process info itself.

rm3l commented on pkg/odo/cli/dev/dev.go:
```diff
@@ -62,9 +62,13 @@ type DevOptions struct {
-	randomPortsFlag bool
-	debugFlag       bool
+	// Experimental behavior as depicted in https://github.com/redhat-developer/odo/issues/5648
+	// If set to false, this should allow to get rid of supervisord and the init container.
```
Thanks for the catch. As discussed above, I'll default to running without Supervisord by default. So I'll remove this flag altogether.

rm3l commented on pkg/remotecmd/interface.go:
```diff
@@ -0,0 +1,51 @@
+
+const (
+	// Unknown represents a process for which the status cannot be determined reliably or is not handled yet by us.
+	Unknown RemoteProcessStatus = iota + 1
```
I was referring to this in the Coding Conventions page : https://github.com/redhat-developer/odo/wiki/Dev:-Coding-Conventions#start-enums-at-one
Because the zero value does have much value here.

dharmit replied:
I was referring to "Learning Go" by Jon Bodner and here's an excerpt from it that made me ask in the first place:
> If there isn't a sensical default value for your constants, a common pattern is to assign the first `iota` value in the constant block to a `_` or to a constant that indicates the value is invalid.

The `Unknown` we have defined here seemed a lot closer to what Jon refers to as "invalid" in the above statement. Anyway, what you pointed to makes sense as well. There's no right or wrong here, IMO. I was only curious, and hence the question. :)

rm3l replied:
Whoops - for some reason, my reply did not get sent. Putting it below for the record :-)

> Interesting - thanks for sharing this. "Unknown" is actually a valid state to me, which we might want to know about. For now, it is coupled with an `error` returned. But while I prefer the `iota + 1` approach (because it reads easily IMO), both recommendations look fine, because the result (starting Enums at one) is the same in the end. I'll add this point to the Coding Conventions wiki page.

rm3l commented on pkg/remotecmd/remotecmd.go:
```diff
@@ -108,13 +130,47 @@ func CreateConsoleOutputWriterAndChannel() (*io.PipeWriter, chan []string) {
+			line := scanner.Text()
+			lines = append(lines, line)
+
+			if log.IsDebug() || show {
```
Yes, I think you're right, like with most languages, `&&` and `||` are short-circuit operators. I guess Go is no exception here. This code was moved from another place, but I'll change this condition.

dharmit commented on pkg/remotecmd/remotecmd.go:
```diff
@@ -8,96 +8,86 @@ import (
+	show bool,
+	consoleOutputStdout *io.PipeWriter,
+	consoleOutputStderr *io.PipeWriter,
+) (stdout []string, stderr []string, err error) {
```
I see two places where `ExecuteCommand` is called. Both places call it like:
```go
_, _, err := ExecuteCommand(...)
```
I wonder if the original signature made sense as far as the return arguments are concerned. WDYT?

dharmit commented on pkg/devfile/adapters/kubernetes/component/commandhandler.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+const numberOfLinesToOutputLog = 100
+const processStatusWaitTimeSeconds = 1
```
Please use `1*time.Second` here instead of converting `int` value to time value. It's more readable and less error-prone, IMO.

rm3l commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,180 @@
+)
+
+const _sigtermExitCode = 143
+const _numberOfLinesToOutputLog = 100
```
When the command does not succeed, `odo dev` displays the last `_numberOfLinesToOutputLog` lines of log. This was the behavior in the previous implementation (i.e., with Supervisord), as tested in [cmd_dev_test.go](https://github.com/redhat-developer/odo/blob/main/tests/integration/devfile/cmd_dev_test.go#L1041-L1062)

rm3l commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,180 @@
+
+// kubeExecProcessHandler implements RemoteProcessHandler by executing Devfile commands right away in the container
+// (like a 'kubectl exec'-like approach). Command execution is done in the background, in a separate goroutine.
+// It works by storing the parent process PID in a file (_startCmdProcessPidFile) in the container,
+// then fires the exec command in the background.
+// The goroutine started can then be stopped by killing the process stored in the state file (_startCmdProcessPidFile)
+// in the container.
```
If I remember correctly from the preliminary tests that we did with Philippe (using `kubectl exec`), stopping the command from the client side sometimes left the process running in the remote container, for some reason.
We did it this way to ensure the remote process is really stopped in the container (while also stopping the goroutine).

feloy commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,192 @@
+	containerName string,
+) (stdout []string, stderr []string, err error) {
+	return Execute(kclient, podName, containerName, false,
+		common.ShellExecutable, "-c", fmt.Sprintf("(kill $(cat %[1]s) || true) && rm -f %[1]s", getPidFileForCommand(devfileCmd)))
```
I think this is equivalent
```
kill $(cat %[1]s); rm -f %[1]s
```

rm3l replied:
Yes, you are both right. This is indeed the expected behavior.
Initially, I added a boolean allowing callers to specify if they want to ignore the exit status of the `kill` command, but reverted it to keep it simple.
I'll use the semi-colon in this case.

feloy commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -69,6 +87,211 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
+}
+
+// getCommandFromDevfile iterates through the devfile commands and returns the command associated with the group
+func getCommandFromDevfile(data data.DevfileData, groupType v1alpha2.CommandGroupKind) (supportedCommand v1alpha2.Command, err error) {
```
I'm not sure about not having  named return values. 
I remember we discussed not to have implcit return statements (a return statement without indicating the values to return)

dharmit replied:
> I remember we discussed not to have implcit return statements (a return statement without indicating the values to return)

Implicit return statement is a result of having a named return values.

feloy replied:
Having named return values is necessary to make implicit return, but you can have named return values and use explicit return.

IMO, named return values adds information to the signature and it would be nice to use it when having some bool or other non auto-documented value in the return signature

dharmit replied:
> Having named return values is necessary to make implicit return, but you can have named return values and use explicit return.

Then this is what I am seeking.

dharmit commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,192 @@
+}
+
+// getPidFileForCommand returns the path to the PID file in the remote container.
+// The parent folder is supposed to be existing.
```
Not figured out so far. But it's working fine, so not going to get crazy over it. 😄

rm3l replied:
Actually, your comment made me think that instead of hardcoding the `/opt/odo` parent path here, I would use the constant already defined in the storage package :)

dharmit replied:
Yes, I remember coming across a constant with that path value. :)

rm3l commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -69,6 +87,211 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
 	return groupCmds[0], nil
 }
 
+// ValidateAndGetPushDevfileCommands validates the build and the run command,
+// if provided through odo push or else checks the devfile for devBuild and devRun.
+// It returns the build and run commands if its validated successfully, error otherwise.
+func ValidateAndGetPushDevfileCommands(
```
Makes sense - I'll rename these functions.

rm3l commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,192 @@
+	// deal with environment variables
+	cmdLine := devfileCmd.Exec.CommandLine
+	setEnvVariable := util.GetCommandStringFromEnvs(devfileCmd.Exec.Env)
+
```
No problem :p Agree with this.

rm3l commented on pkg/remotecmd/remotecmd.go:
```diff
@@ -8,96 +8,86 @@ import (
+	_, _, err = ExecuteCommand(kclient, containerName, podName, cmd, show, stdoutWriter, stderrWriter)
+	if err != nil {
+		klog.V(2).Infof("unable to execute command on %s within container %s, %v, output: %v %v: %+v",
+			podName, containerName, err, stdout, stderr, err)
```
Thanks for spotting this. No, this was not intentional. I'll update it.

dharmit replied:
Ah, so there's some benefit of reading line by line. 😉  (kidding...)

dharmit commented on pkg/machineoutput/event_logging.go:
```diff
@@ -232,35 +216,35 @@ func (c *ConsoleMachineEventLoggingClient) outputJSON(machineOutput MachineEvent
 // Currently used for test purposes only.
 func (w MachineEventWrapper) GetEntry() (MachineEventLogEntry, error) {
-
 	if w.DevFileCommandExecutionBegin != nil {
```
Yep, it's indeed more readable. And I understand that a switch needs to have some relationship (maybe even matching the types). But I didn't look in detail, and just blurted out my thoughts. :)

dharmit commented:
/lgtm

dharmit commented on pkg/remotecmd/remotecmd.go:
```diff
@@ -108,13 +95,47 @@ func CreateConsoleOutputWriterAndChannel() (*io.PipeWriter, chan []string) {
+				}
+			}
+		}
+		result <- lines
```
I saw this one coming. 😉

rm3l commented on pkg/remotecmd/interface.go:
```diff
@@ -0,0 +1,59 @@
+	Status RemoteProcessStatus
+}
+
+// RemoteProcessHandler is an interface for managing processes that are intended to be executed remotely,
+// in Kubernetes/OpenShift containers
+type RemoteProcessHandler interface {
```
I would say both ^^. Its purpose is to execute commands remotely and handle the remote processes linked to such commands.
But as I generally struggle with naming things, I am open to a different name if you have a better suggestion.

dharmit replied:
Nope, it's just fine.

rm3l commented on pkg/sync/adapter_test.go:
```diff
@@ -97,16 +99,17 @@ func TestSyncFiles(t *testing.T) {
 
 	ctrl := gomock.NewController(t)
+	kc := kclient.NewMockClientInterface(ctrl)
+	kc.EXPECT().ExecCMDInContainer(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
```
I was looking at your test in your PR. Not an expert of gomock, but to some extent, yes, I think you could have used `gomock.Any()` to mock interactions with `o.kubernetesClient`. However, I think this would be of limited help given the signature of the `matchOwnerReferenceWithResources` method you are testing: you would still need to construct and pass `resources []unstructured.Unstructured` to it, which seems to be the point of your `fakePod` and `fakeDeployment` functions.

feloy commented on pkg/devfile/adapters/kubernetes/utils/utils.go:
```diff
@@ -186,8 +143,8 @@ func UpdateContainersWithSupervisord(devfileObj devfileParser.DevfileObj, contai
 			} else {
-				// if the env has the debug port but if its value is different then we change it
-				// this way the value of env info is given more importance then whats there on the devfile
+				// if the env has the debug port but if its value is different, then we change it.
```
Yes, I cannot see anywhere this value written in this file

rm3l commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,192 @@
+		return process, err
+	}
+
+	if len(stdout) == 0 {
```
Wanted to be safe before accessing the first element of `stdout` ^^, but you're right, `stdout` can never be empty here. I blindly copied what I wrote in `getRemoteProcessPID`.

rm3l commented:
@feloy @dharmit I've addressed your comments and also pushed a few changes:

- as we saw together with @feloy last week, getting and stopping the children processes now solves the issue of stopping the remote process associated with the Devfile command
- rather than waiting for a few seconds right after the remote process associated with the Devfile command is stopped (as previously done with the Supervisord mode), we're now retrying to check the process status (until a little timeout), to make sure it is really stopped before starting it again. This is the purpose of the `pkg/task` package and the functions in `retry.go`. Otherwise, there were some situations where the process would take slightly longer to stop.
- we are now able to make it able to return the Devfile command process logs via `odo logs` or `kubectl logs`. I also added a bunch of unit tests for both `kubeexec.go` and `remotecmd.go`.

I've tested these against a few Devfiles, to make sure everything works as expected.

That would be great if you (or anyone else) could take another look. Thanks.

rm3l commented:
/test v4.10-integration-e2e

rm3l commented:
I just rebased the branch to benefit from `odo logs`, now that #5760 has been merged.

rm3l commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -69,6 +87,214 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
+}
+
+// getCommandAssociatedToGroup iterates through the devfile commands and returns the command associated with the group
+func getCommandAssociatedToGroup(data data.DevfileData, groupType v1alpha2.CommandGroupKind) (v1alpha2.Command, error) {
```
As I understand it, `getCommandAssociatedToGroup` either returns the default command for this group or the first non-default one it found.
On the other hand, `GetDefaultCommand` returns an error if there is no default command for the specified group. It also checks that there is not more than one default command found for this group while iterating.
So they seem to behave differently. But I agree that they look similar.

feloy replied:
I think we should try to haromize the behaviour between commands. Deploy is using `getDefaultCommand`, when Build/Test/Run is using `getCommandAssociatedToGroup`

rm3l replied:
Yes, makes sense. I had the same feeling when moving all those functions to `libdevfile`, but wanted to refactor them when working on supporting non-default build and run commands (either #5775 or #5776), since these functions will need to be updated for these issues. WDYT?

feloy replied:
Yes, makes sense to work on it during #5775 or #5776

rm3l replied:
Okay, I added a comment to these issues to keep track of this point. Thanks.

feloy commented:
/approve

Great work on refactoring this part of the code. Thanks @rm3l

feloy commented:
> /hold I'm seeing some weird behavior especially with java-springboot.
> 
> with old approach I don't remember last time `odo dev` failed on me. But when testing this PR i got following error twice
> 
> ```
> ▶ odo dev
>   __

>  ✗  Waiting for Kubernetes resources [1m]

What is the total deployment time for other sessions that don't fail? It seems to me that 1m is pretty short to - delete the Container image - sync the files - build the binary - start the program

valaparthvi commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,294 @@
+	// Change to the workdir and execute the command
+	pidFile := getPidFileForCommand(devfileCmd)
+	cmd := []string{common.ShellExecutable, "-c"}
+	// Storing the /bin/sh parent process PID. It will allow to determine its children later on and kill them when a stop is request
```
```suggestion
	// Storing the /bin/sh parent process PID. It will allow to determine its children later on and kill them when a stop is requested
```

valaparthvi commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,294 @@
+	// Redirecting to /proc/1/fd/* allows to redirect the process output to the output streams of PID 1 process inside the container.
+	// This way, returning the container logs with 'odo logs' or 'kubectl logs' would work seamlessly.
+	// See https://stackoverflow.com/questions/58716574/where-exactly-do-the-logs-of-kubernetes-pods-come-from-at-the-container-level
+	outputRedirectCmd := "1>>/proc/1/fd/1 2>>/proc/1/fd/2"
```
> Full command is: /bin/sh -c "echo $$ > $pidFile && cd $workingDir && ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2"

Can you explain the ` ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2` part of the cmd? I understand the redirection, but I'm not sure what this command is supposed to do?

rm3l replied:
`$cmdline` is the Devfile run or debug command we are supposed to run. The parentheses around it mean that we run this command in a subshell, which will be the child of the `/bin/sh` process.
This was done to make sure we could handle cases where the command-line string defined in the Devfile does more things, like `sleep 10; <do_something_in_background> & <start_my_server>`.

Finally, the redirection is to have the output of this subshell process redirected to the container main (PID 1) process, so `odo logs` or `kubectl logs` could return such logs as well.

valaparthvi commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,294 @@
+		}
+	}()
+
+	ppid, err := getRemoteProcessPID(kclient, devfileCmd, podName, containerName)
+	if err != nil {
+		return err
+	}
+	if ppid == 0 {
+		return nil
+	}
+
+	children, err := getProcessChildren(ppid, kclient, podName, containerName)
+	if err != nil {
+		return err
+	}
```
Nitpick: Since these are not used until after `kill` is defined, I would move their position to be placed after `kill`.

valaparthvi commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,294 @@
+
+// StopProcessForCommand stops the process representing the specified Devfile command.
+// Because of the way this process is launched and its PID stored (see StartProcessForCommand),
+// we need to determine the process children (there should be only one child). Then killing those children
```
Is there a reason why there should only be one child?

rm3l replied:
The Devfile command is executed in a subshell, which is the child here (this is the intent of the `($cmdLine)` you asked about previously).
This was done to make sure we could handle cases where the command-line string defined in the Devfile does more things, like 'sleep 10; <do_something_in_background> & <start_my_server>'

valaparthvi commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,294 @@
+		return nil
+	}
+
+	if len(children) == 0 {
+		return kill(ppid)
+	}
```
I don't see why you need to do a separate check for `len(children)`, I think using the for loop instead would be fine and not make much difference. Also, if there should only be one child, shouldn't we check that here and raise an error if it's not true?

rm3l replied:
> I don't see why you need to do a separate check for `len(children)`, I think using the for loop instead would be fine and not make much difference. 
>
I wanted to fall back to killing the parent process if there were no identified children. But you're right, this could have been done with a single loop iterating over both the parent and all the children PIDs found. I'll change that.

> Also, if there should only be one child, shouldn't we check that here and raise an error if it's not true?
>
For now, there is one child, but I wanted to be safe if we later change the final command-line and end up with several children. It was cheap to add, so I wanted to be able to handle that as well.

valaparthvi commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -69,6 +87,214 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
 }
 
+// ValidateAndGetPushCommands validates the build and the run command,
+// if provided through odo push or else checks the devfile for devBuild and devRun.
```
```suggestion
// if provided through odo dev or else checks the devfile for devBuild and devRun.
```

valaparthvi commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -69,6 +87,214 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
+
+	isBuildCmdEmpty := reflect.DeepEqual(emptyCommand, buildCommand)
+	if isBuildCmdEmpty && buildCmdErr == nil {
+		// If there was no build command specified through odo push and no default build command in the devfile, default validate to true since the build command is optional
```
```suggestion
		// If there was no build command specified through odo dev and no default build command in the devfile, default validate to true since the build command is optional
```

valaparthvi commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -69,6 +87,214 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
+// ValidateAndGetPushCommands validates the build and the run command,
+// if provided through odo push or else checks the devfile for devBuild and devRun.
+// It returns the build and run commands if its validated successfully, error otherwise.
+func ValidateAndGetPushCommands(
```
Do you think we can use `ValidateCommands` function from the devfile library instead of defining separate functions for every command type?
https://github.com/redhat-developer/odo/blob/aa2d935e7f4e7da2bf7249458a5abb9baca2605e/vendor/github.com/devfile/api/v2/pkg/validation/commands.go#L15

rm3l replied:
When looking at the call tree, this `ValidateCommands` seems to be called indirectly when we call `parseDevfile` in the `pkg/devfile` package. So not sure about this, but I intend to have a closer look at simplifying this `libdevfile.go` file when working on either #5775 or #5776.

rm3l commented:
> /hold I'm seeing some weird behavior especially with java-springboot.
> 
> with old approach I don't remember last time `odo dev` failed on me. But when testing this PR i got following error twice
> 
> ```
> ▶ odo dev
>   __
>  /  \__     Developing using the my-java-springboot-app Devfile
>  \__/  \    Namespace: demo
>  /  \__/    odo version: v3.0.0-alpha3
>  \__/
> 
> ↪ Deploying to the cluster in developer mode
>  ✓  Added storage m2 to my-java-springboot-app
>  ✓  Searching resource in cluster es
>  ✓  Creating kind ServiceBinding [58ms]
>  ✓  Searching resource in cluster 
>  ✓  Creating kind ServiceBinding [21ms]
>  ◐  Waiting for Kubernetes resources^C
> 
> Cancelling deployment.
> This is non-preemptive operation, it will wait for other tasks to finish first
> 
>  ✗  Waiting for Kubernetes resources [1m]
> Cleaning resources, please wait
> Failed to delete the "Deployment" resource: my-java-springboot-app-app
>  ✗  failed to create the component: timeout while waiting for "my-java-springboot-app-app-747d44d4f7-hm8rq" pod to be deleted
> ```
> 
> I'm trying to debug it but so far I was not able to reproduce it with verbose logs enabled.
> 
> /hold

Did this happen right after running `odo dev` or later after a file change?
I remember I had a similar issue with ServiceBinding, with odo waiting for some resources, due to #5781, I think. Anyways, I have just rebased this branch onto `main`. I tested it again with a SpringBoot application and a binding in the Devfile.
Could you check again with this version?

rm3l commented:
> > /hold I'm seeing some weird behavior especially with java-springboot.
> > with old approach I don't remember last time `odo dev` failed on me. But when testing this PR i got following error twice
> > ```
> > ▶ odo dev
> >   __
> > ```
> 
> > ✗  Waiting for Kubernetes resources [1m]
> 
> What is the total deployment time for other sessions that don't fail? It seems to me that 1m is pretty short. I remember getting this error when the image is too long to download

Is this timeout hardcoded in odo?

feloy commented:
> > > /hold I'm seeing some weird behavior especially with java-springboot.
> > > with old approach I don't remember last time `odo dev` failed on me. But when testing this PR i got following error twice
> > > ```
> > > ▶ odo dev
> > >   __
> > > ```
> > 
> > 
> > > ✗  Waiting for Kubernetes resources [1m]
> > 
> > 
> > What is the total deployment time for other sessions that don't fail? It seems to me that 1m is pretty short. I remember getting this error when the image is too long to download
> 
> Is this timeout hardcoded in odo?

I think it is the PushTimeout you can set in preferences

rm3l commented:
> > > > /hold I'm seeing some weird behavior especially with java-springboot.
> > > > with old approach I don't remember last time `odo dev` failed on me. But when testing this PR i got following error twice
> > > > ```
> > > > ▶ odo dev
> > > >   __
> > > > ```
> > > 
> > > 
> > > > ✗  Waiting for Kubernetes resources [1m]
> > > 
> > > 
> > > What is the total deployment time for other sessions that don't fail? It seems to me that 1m is pretty short. I remember getting this error when the image is too long to download
> > 
> > 
> > Is this timeout hardcoded in odo?
> 
> I think it is the PushTimeout you can set in preferences

Okay, I see - thanks.

rm3l commented on pkg/remotecmd/interface.go:
```diff
@@ -0,0 +1,66 @@
+// RemoteProcessStatus is an enum type for representing process statuses.
+type RemoteProcessStatus int
+
+const (
+	// Unknown represents a process for which the status cannot be determined reliably or is not handled yet by us.
+	Unknown RemoteProcessStatus = iota + 1
+
+	// Starting represents a process that is just about to start.
+	Starting
+
+	// Stopped represents a process stopped.
+	Stopped
+
+	// Running represents a running process.
+	Running
+)
```
> This bit is interesting, I haven't seen anything like it before! 
> But I am wondering, if changing the order of these constants, thereby changing their values, will have any effect to their usages.
>
The values do not matter here; we refer to the constants by name rather than by value. So changing the order (or even inserting new statuses in the list) will have no effect.
More details in the coding conventions doc: https://github.com/redhat-developer/odo/wiki/Dev:-Coding-Conventions#start-enums-at-one

kadel commented:
> Did this happen right after running `odo dev` or later after a file change?
> I remember I had a similar issue with ServiceBinding, with odo waiting for some resources, due to #5781, I think. Anyways, I have just rebased this branch onto `main`. I tested it again with a SpringBoot application and a binding in the Devfile.
> Could you check again with this version?

It looks like my failure is not related to this PR. 
Now it is consistently failing even on main, or with alpha3 release.

I'm debugging the problem, but it looks like something related to SBO
My Binding has a following status 
```
  status:
    conditions:
    - lastTransitionTime: "2022-06-16T08:04:49Z"
      message: ""
      reason: DataCollected
      status: "True"
      type: CollectionReady
    - lastTransitionTime: "2022-06-16T08:04:49Z"
      message: Unable to retrieve ClusterWorkloadResourceMapping for type "apps/v1,
        Resource=deployments"
      reason: ApplicationNotFound
      status: "False"
      type: InjectionReady
    - lastTransitionTime: "2022-06-16T08:04:49Z"
      message: Unable to retrieve ClusterWorkloadResourceMapping for type "apps/v1,
        Resource=deployments"
      reason: ProcessingError
      status: "False"
      type: Ready
```

Even tho the application is defined correctly 
```
 spec:
    application:
      group: apps
      name: myspring-app
      resource: deployments
      version: v1
```

```
▶ k get deployment myspring-app 
NAME           READY   UP-TO-DATE   AVAILABLE   AGE
myspring-app   1/1     1            1           7s

```
It is still a problem. 
Removing the hold as this is not related to this PR.
/hold cancel

rm3l commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,294 @@
+func getRemoteProcessPID(kclient kclient.ClientInterface, devfileCmd devfilev1.Command, podName string, containerName string) (int, error) {
+	pidFile := getPidFileForCommand(devfileCmd)
+	stdout, stderr, err := ExecuteCommandAndGetOutput(kclient, podName, containerName, false,
+		common.ShellExecutable, "-c", fmt.Sprintf("cat %s || true", pidFile))
```
Yes, it was in case the pid file does not exist, for example the very first time we check if the process is running. I didn't want the command to fail (regardless of the error), because it would make it a bit complicated (for nothing, IMO, as this is not an issue) to additionally check `err` for the type of exit code. What matters most here is the content of the standard output of this `cat` command.

rm3l commented on pkg/remotecmd/remotecmd.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+// ExecuteCommandAndGetOutput executes the given command in the pod's container, and returns the command stdout and stderr content
+func ExecuteCommandAndGetOutput(
```
They used to be slightly different, but you're right; now that both return an output, I've merged those into a single `ExecuteCommand` function.

rm3l commented on pkg/remotecmd/kubeexec.go:
```diff
@@ -0,0 +1,192 @@
+// StartProcessForCommand runs the (potentially never finishing) Devfile command in the background.
+// The goroutine spawned here can get stopped either by stopping the parent process (e.g., 'odo dev'),
+// or by stopping the underlying remote process by calling the StopProcessForCommand method.
+func (k *kubeExecProcessHandler) StartProcessForCommand(
```
Whoops - I missed @dharmit's initial comment. Yes, that's a good point. I've changed the implementation a little bit to remove the dependency to a devfile command. Please take a look at [241302202e2cdc19ab4142511760c79ebd2e404b](https://github.com/redhat-developer/odo/pull/5768/commits/241302202e2cdc19ab4142511760c79ebd2e404b)

valaparthvi reviewed (approved):
The code lgtm, thank you for patiently working on this!

feloy commented:
/override ci/prow/unit
/override ci/prow/v4.10-integration-e2e
